authority/provisioner/azure.go

```diff
@@ -0,0 +1,299 @@
+}
+
+type azurePayload struct {
+	jose.Claims
+	AppID            string `json:"appid"`
+	AppIDAcr         string `json:"appidacr"`
+	IdentityProvider string `json:"idp"`
+	ObjectID         string `json:"oid"`
+	TenantID         string `json:"tid"`
+	Version          string `json:"ver"`
+	XMSMirID         string `json:"xms_mirid"`
+}
+
+// Azure is the provisioner that supports identity tokens created from the
+// Microsoft Azure Instance Metadata service.
+//
+// The default audience is "https://management.azure.com/".
+//
+// If DisableCustomSANs is true, only the internal DNS and IP will be added as a
+// SAN. By default it will accept any SAN in the CSR.
+//
+// If DisableTrustOnFirstUse is true, multiple sign request for this provisioner
+// with the same instance will be accepted. By default only the first request
+// will be accepted.
+type Azure struct {
+	Type                   string   `json:"type"`
+	Name                   string   `json:"name"`
+	TenantID               string   `json:"tenantId"`
+	ResourceGroups         []string `json:"resourceGroups"`
+	Audience               string   `json:"audience,omitempty"`
+	DisableCustomSANs      bool     `json:"disableCustomSANs"`
+	DisableTrustOnFirstUse bool     `json:"disableTrustOnFirstUse"`
+	Claims                 *Claims  `json:"claims,omitempty"`
+	claimer                *Claimer
+	config                 *azureConfig
+	oidcConfig             openIDConfiguration
+	keyStore               *keyStore
+}
+
+// GetID returns the provisioner unique identifier.
+func (p *Azure) GetID() string {
+	return p.TenantID
+}
+
+// GetTokenID returns the identifier of the token. The default value for Azure
+// the SHA256 of "xms_mirid", but if DisableTrustOnFirstUse is set to true, then
+// it will be the token kid.
+func (p *Azure) GetTokenID(token string) (string, error) {
+	jwt, err := jose.ParseSigned(token)
+	if err != nil {
+		return "", errors.Wrap(err, "error parsing token")
+	}
+
+	// Get claims w/out verification. We need to look up the provisioner
+	// key in order to verify the claims and we need the issuer from the claims
+	// before we can look up the provisioner.
+	var claims azurePayload
+	if err = jwt.UnsafeClaimsWithoutVerification(&claims); err != nil {
+		return "", errors.Wrap(err, "error verifying claims")
+	}
+
+	// If TOFU is disabled create return the token kid
+	if p.DisableTrustOnFirstUse {
+		return claims.ID, nil
+	}
+
+	sum := sha256.Sum256([]byte(claims.XMSMirID))
+	return strings.ToLower(hex.EncodeToString(sum[:])), nil
+}
+
+// GetName returns the name of the provisioner.
+func (p *Azure) GetName() string {
+	return p.Name
+}
+
+// GetType returns the type of provisioner.
+func (p *Azure) GetType() Type {
+	return TypeAzure
+}
+
+// GetEncryptedKey is not available in an Azure provisioner.
+func (p *Azure) GetEncryptedKey() (kid string, key string, ok bool) {
+	return "", "", false
+}
+
+// GetIdentityToken retrieves the identity document and it's signature and
```

Comment:
no identity doc / signature involved here. Just getting the token, as far as I can tell.

Reply:
Fixed with cf07c8f